lighthouse-viewer/app/src/main.js

```diff
@@ -0,0 +1,504 @@
+    });
+
+    document.addEventListener('drop', e => {
+      e.stopPropagation();
+      e.preventDefault();
+
+      this._resetDraggingUI();
+
+      // Note, this ignores multiple files in the drop, only taking the first.
+      this._fileHandlerCallback(e.dataTransfer.files[0]);
+    });
+  }
+
+  _resetDraggingUI() {
+    this.dropZone.classList.remove('dropping');
+    this._dragging = false;
+    logger.hide();
+  }
+
+  removeDropzonePlaceholder() {
+    // Remove placeholder drop area after viewing results for first time.
+    // General dropzone takes over.
+    if (this.placeholder) {
+      this.placeholder.remove();
+      this.placeholder = null;
+    }
+  }
+
+  /**
+   * Reads a file and returns its content in the specified format.
+   * @static
+   * @param {!File} file
+   * @param {!string} readAs A format to read the file ('text', 'dataurl',
+   *     'arraybuffer', 'binstr'). The default is to return the file as text.
+   * @return {!Promise<string>}
+   */
+  static readFile(file, readAs) {
+    return new Promise((resolve, reject) => {
+      const reader = new window.FileReader();
+      reader.onload = function(e) {
+        resolve(e.target.result);
+      };
+      reader.onerror = reject;
+      switch (readAs) {
+        case 'arraybuffer':
+          reader.readAsArrayBuffer(file);
+          break;
+        case 'binstr':
+          reader.readAsBinaryString(file);
+          break;
+        case 'dataurl':
+          reader.readAsDataURL(file);
+          break;
+        case 'text':
+        default:
+          reader.readAsText(file);
+      }
+    });
+  }
+}
+
+class LighthouseViewerReport {
+  constructor() {
+    this.onShare = this.onShare.bind(this);
+    this.onFileUpload = this.onFileUpload.bind(this);
+
+    this.json = null;
+    this.fileUploader = new FileUploader(this.onFileUpload);
+    this.github = new GithubAPI();
+
+    this.addListeners();
+    this.loadFromURL();
+  }
+
+  addListeners() {
+    const printButton = document.querySelector('.js-print');
+    if (!printButton) {
+      return;
+    }
+
+    printButton.addEventListener('click', _ => {
+      window.print();
+    });
+
+    const button = document.createElement('button');
+    button.classList.add('share');
+    button.addEventListener('click', this.onShare);
+    printButton.parentElement.insertBefore(button, printButton);
+  }
+
+  loadFromURL() {
+    // Pull gist id from URL and render it.
+    const params = new URLSearchParams(location.search);
+    const gistId = params.get('gist');
+    if (gistId) {
+      logger.log('Fetching report from Github...', false);
+
+      this.github.auth.ready.then(_ => {
+        this.github.getGistContent(gistId).then(json => {
+          logger.hide();
+          this.replaceReportHTML(json.content);
+
+          // Save fetched json and etag to IDB so we can retrieve use it later
+          // for 304 requests.
+          return idb.set(gistId, {etag: json.etag, content: json.content});
```

Comment:
Right now, `getGistContent` doesn't check if the JSON returned int he gist is valid lighthouse json. That's checked later in `replaceReportHTML`, which will throw if it's unknown json.

Added a comment.